Retrieve the code block content from the odoc-parser locations

Leonidas-from-XIV opened the pull request:
The reason for this is due to us wanting to have the spaces available in MDX, so we know all the whitespace, so we can format it the way it came in from the user.

This also addresses the parsing issue for multiline toplevel blocks.

Closes #394

panglesd commented on lib/block.ml:
```diff
@@ -145,41 +144,44 @@ let dump ppf ({ loc; section; labels; contents; value; _ } as b) =
-    | Some Syntax.Mli ->
-        fun ppf -> Fmt.fmt "%*s%s" ppf (t.loc.loc_start.pos_cnum + 2) ""
-    | _ -> Fmt.string
+let pp_lines syntax _t ppf lines =
+  let print_line ppf first last line =
+    (match syntax with
+    | Some Syntax.Cram -> Fmt.string ppf line
+    | Some Syntax.Mli -> (
+        match (first, last) with
+        | true, _ | _, true -> Fmt.string ppf (String.trim line)
+        | false, false -> Fmt.string ppf line)
+    | Some Syntax.Normal | None -> Fmt.string ppf line);
+    match last with false -> Fmt.pf ppf "\n" | true -> ()
```
I am not sure why this trimming on first and last lines in mli files. in particular, with this:
```
{[             
   "11" ]}
```
is tranformed like this (there are removed whitespace on the first line):
```
-{[             
-   "11" ]}
+{[
+"11"]}
```

I suggest:
```suggestion
  let print_line ppf first last line =
    Fmt.string ppf line;
    match last with false -> Fmt.pf ppf "\n" | true -> ()
```
which however require changes elsewhere as it breaks the tests: something must be adding padding for last line!

panglesd commented on lib/mli_parser.ml:
```diff
@@ -53,18 +56,68 @@ let slice lines ~(start : Odoc_parser.Loc.point) ~(end_ : Odoc_parser.Loc.point)
+  (* first line always starts at index zero, even if there is no preceeding newline *)
+  let indexes = 0 :: List.of_seq indexes_of_line_starts in
+  fun nth ->
+    (* index starts at zero but lines go from 1 *)
```
Good job at figuring all these out!

Let's hope `odoc-parser` does not change those facts…

panglesd reviewed (approved):
Thanks @Leonidas-from-XIV for the PR!

I left some minor comments in the code.

Having the comments formatted the way they came is very useful for compatibility with code formatters. (There are still requirements on the format for toplevel blocks and cram blocks, so the code formatter should not do too invasive changes to keep compatibility with mdx).

Note that currently it is not exactly formatted the way it comes from the user: there are cases related to the formatting of last and first lines which changes this input. However, they are strange formatting and I think it is safe to modify them (if you manage to get a true preservation of input that's even better!). See https://github.com/realworldocaml/mdx/pull/397#discussion_r987640954

It's very nice to see all those `match syntax with ...` be removed!

Leonidas-from-XIV commented on CHANGES.md:
```diff
@@ -19,6 +19,8 @@
 - Include the content of the line that features the `part-end` MDX directive in
   the output, before that line would've been dropped (#374, #387,
   @Leonidas-from-XIV)
+- Fix parsing of multiline toplevel phrases in .mli files (#394, #397,
+  @Leonidas-from-XIV)
```
They sort of are, but this PR does a lot of "under the table reshuffling" so the description in the changelog is essentially just what made me start this endevour. It sort of also preserves the indentation, like #395, but refactors the way this is done.

Another way to describe this could be:

```
- Preserve indentation in multiline OCaml blocks in .mli files and fix multiline cram blocks (#395, #397, @panglesd, @Leonidas-from-XIV)
```

But I feel like one of them is a change and one is pretty much a bugfix, so joining them feels a bit unfortunate.

The problem after working with a PR for too long of a time one looses the user perspective and sees the technical change, but the Changelog is specifically written for users. If you have a suggestion, I'll gladly incorporate it!

Alnsim69 replied:
اريد الدخول الى حسابي الخاص فيسبوك

Leonidas-from-XIV commented on lib/lexer_mdx.mll:
```diff
@@ -64,12 +64,12 @@ and cram_text section = parse
         let header = "sh" in
         let requires_empty_line, contents = cram_block lexbuf in
-        let contents = first_line :: contents in
+        let contents = (Format.asprintf "%s%s" ws first_line) :: contents in
```
This is because by matching on `ws` in the lexer rule it eats the whitespace, so `first_line` doesn't contain the leading whitespace. An important aspect of this refactoring would be to preserve the whitespace, so I added the whitespace back on.

A problematic aspect is that due to the lexer in MDX files the `cram` blocks have to be indented by exactly two spaces, so indenting by four wouldn't be accepted but I deemed this an aspect that can be improved at a later point in time, the rest of the code should be able to work with any amount of `hpad` as long as it is consistent.

Leonidas-from-XIV commented on lib/cram.ml:
```diff
@@ -67,38 +52,40 @@ let pp ?pad ppf (t : t) =
+        | false -> String.with_index_range line ~first:hpad)
   in
   let lines = List.map unpad t in
+  (* remove leading empty line, if any which is true for multipline blocks *)
+  let lines = match lines with "" :: lines -> lines | lines -> lines in
```
This code is only used to remove the initial empty line that is created in multiline blocks (since the first line starts right at the beginning of a block and usually is just followed by a line break before the "proper" indented block starts (which is also why `hpad_of_lines` recurses, because the first line might be an outlier and not have any `hpad` at all)).

Now that you point it out, it is somewhat of an ugly hack because the cram parser rejects it otherwise but I feel like this might be the wrong approach. It seems to lead to the issue https://github.com/realworldocaml/mdx/pull/397/files#diff-406c39c844335a646d3100b6fba8ae3b2add673571869dd2a7315ba00b3e03f9R100 where a `\n` needs to be conditionally added because we removed it here and lost the information whether it was there or not.

Maybe we should just not mess with the empty lines and preserve them? What do you think?

panglesd replied:
I would say that the least we have to have special cases with blank lines, the best.
So, I think it is better to keep the blank lines, and modify `lexer_cram.mll` to ignore leading whitespace. Not sure though!

```suggestion
```

Leonidas-from-XIV replied:
It's a good point. Upon investigation the tokenizer just parses those as `` `Output ""`` so I wrote some code to remove them and convert them to `vpad`. Then I also removed some more special casing where depending on whether it is in a cram or `.mli` file it would output empty outputs in a padded or non-padded way. It's somewhat of an edge-case but I think being consistent and predictable trumps avoiding trailing spaces.

Leonidas-from-XIV commented on test/bin/mdx-test/expect/multilines-mli/test-case.mli:
```diff
@@ -33,4 +33,15 @@ The formatting for multilines in .mli files is the following:
       | None -> ()
       | Some b -> b
   ]}
+
+It should also work fine for toplevel descriptions (as in
+[multilines/test-case.md]):
+
+{@ocaml[
+  # let rec fact = function
```
Good catch, I've updated it!

Leonidas-from-XIV commented:
@panglesd Can you take a look again at the PR? I've decided to refactor the way tests are handled by not removing the padding which triggered a lot of changes down the line, since most formats were printing their beginning/end-markers except for MLI so in my quest to unify the way it they get printed I had to revisit a significant amount of code since it needed to be made consistent.

This took a quite significant amount of time because there was a lot of subtlety how everything is printed, but in the end I thing the result makes sense and should be easier to maintain.

I've also refactored some odd side-effectful code and split the `Misc` module up a bit to be less of a grab-bag of functions and have a little bit more focus.

Unfortunately, that's quite a bit of changes added on top, but once I started digging I wanted to finish what I started.

panglesd commented:
Sure, I'll have a look ASAP! (ASAP might be next week)

panglesd reviewed (approved):
It is hard to go through such big refactoring!

So I cannot say I have verified every modification. I have some comments, but it looks a nice improvement!

First, from my testing, everything work well and makes it working well with automatic formatter. About empty line preservation, there are some non-important "inconsistencies", that I report here:
- In toplevel blocks, empty lines are preserved when positioned first, in between blocks, but not last. For instance:
```ocaml
  {@ocaml[

    # print_endline "foo";;
    foo
    - : unit = ()

    # print_endline "bar";;
    bar
    - : unit = ()

  ]}
```
will be rewritten to:
```ocaml
  {@ocaml[

    # print_endline "foo";;
    foo
    - : unit = ()

    # print_endline "bar";;
    bar
    - : unit = ()
  ]}
```
- In shell, lines are preserved at first, but not inbetween, or last:
```ocaml
  {@sh[

    $ echo foo
    foo

    $ echo bar
    bar

  ]}
```
will be rewritten to:
```ocaml
  {@sh[

    $ echo foo
    foo
    $ echo bar
    bar
  ]}
```
- Normal fragments work as expected: nothing is changed from the input.

About the code, it looks much better without all those `?syntax` everywhere. I have some comments:
- I think it would be better not to have a `Language_tag` variant in the `label` type. It does not enforce the fact that there can be only one language tag (while it is enforced by the syntax).
- There are more `Log.debug` than before. I don't know if it's intended, or a debug trace that should be removed...
- Mli syntax is much less tested than markdown's one. For instance, labels parsing and reconstructing should be tested (and made less constrained). Although I don't know how much this is related to this PR (which was I think started to improve mli syntax parsing and reconstruction).

Leonidas-from-XIV commented on lib/label.ml:
```diff
@@ -85,6 +85,7 @@ type t =
   | Set of string * string
   | Unset of string
   | Block_kind of block_kind
+  | Language_tag of string
```
I agree and my first approach was to use `block_kind`/`syntax` for it, but the issue is that this doesn't preserve whether the language tag existed when it was parsed, thus it would add tags where the language was the same as automatically determined (creating diff noise) or if we implemented it to only emit it if it is not the default, it would remove the tags where the input had the default language (creating diff noise again). Thus I would argue that we *need* some place to track this.

It is true that `Language_tag` should only exist at most once, so a nicer way would be a `string option`. Though the same is true for `Block_kind`, and possibly other labels like `Version`. It seems `Label.t list` (as used in `Block.t`) is not a good way to express the labels, `Label.t` should probably be a record with `'a option` fields for many of the tags (in fact, only `Set`, `Unset` and `Env` seem like they should be lists).

Do you think it makes sense to refactor this before merging? I could do this, but part of the reason why this PR ended being so big is that I came across a lot of bits that needed drive-by refactoring so I was a bit hesistant on changing even more (especially as that might change the semantics and reject previously valid files for multiple tags). For practical reasons I think deferring this to a future PR might be better (partly for the fact that it would get a clear changelog entry).

Leonidas-from-XIV replied:
That's a good point, I added an explanation of the tag and a refactoring suggestion (and rebased to fix the changelog).

Leonidas-from-XIV commented on lib/toplevel.ml:
```diff
@@ -76,43 +95,64 @@ let lexbuf ~(pos : Lexing.position) s =
+          !i)
+
+let of_lines ~(loc : Location.t) t =
+  Log.debug (fun l -> l "Toplevel lines to parse %a" Fmt.Dump.(list string) t);
```
Good point. I added a lot of debug prints when working on this, but most can indeed be removed.

Leonidas-from-XIV commented on lib/block.ml:
```diff
@@ -422,9 +408,13 @@ let mk ~loc ~section ~labels ~legacy_labels ~header ~contents ~errors =
+  Logs.debug (fun l ->
+      l "Block kind in Block.mk is %a"
+        (Fmt.Dump.option Label.pp_block_kind)
+        block_kind);
```
Nope, just leftoever debug prints. Thanks for catching!

Leonidas-from-XIV commented on lib/cram.mli:
```diff
@@ -20,8 +20,17 @@ type t = {
   command : string list;
   output : Output.t list;
   exit_code : int;
+  (* TODO: remove vpad *)
```
Good point. I had it set to 0 temporarily, since there's no more padding between cram tests (the padding is now part of another type), but indeed, now that I'm done I can remove it as it is not coming back.

Leonidas-from-XIV commented on lib/test/mdx_test.ml:
```diff
@@ -300,36 +349,30 @@ let run_exn ~non_deterministic ~silent_eval ~record_backtrace ~syntax ~silent
-          in
-          with_non_det non_deterministic non_det ~command:print_block
-            ~output:(fun () ->
+          (* t.contents for md files is wrong, it's missing the begin and end *)
```
I'm not quite clear about it myself anymore.

I think the issue is that in markdown `t.contents` is missing the initial newline (and potential trailing newline, if it doesn't end the file), so there's another place where the parsers for the various formats have slightly different behavior.

I think this is another hole to dig and can be tackled at some other point in time. The comment is quite unclear so to prevent people from wondering what this cryptic scribble means I'll remove it.

Leonidas-from-XIV commented on lib/block.ml:
```diff
@@ -145,77 +145,86 @@ let dump ppf ({ loc; section; labels; contents; value; _ } as b) =
-let pp_header ?syntax ppf t =
+let pp_labels ?syntax ppf labels =
   match syntax with
+  | Some Syntax.Mli -> Fmt.(list ~sep:(any ",") Label.pp) ppf labels
```
Agreed. I've added a test for the labels syntax by copying `labels-syntax` and making it an `mli` file. It worked out of the box without issues so that's good.

The disagreements on supported syntax is quite annoying between the various parsers, so over time we should move the `mli` parser closer to the other parsers.

Leonidas-from-XIV commented:
I'll merge this now since I think it is a decent point and further improvements can be added in future (hopefully shorter) PRs. Thanks for the review @panglesd!
